### ISSUE TITLE
THIS REPO IS ABADONED >>> github.com/sannidhyaroy/Soduto has taken over

### ISSUE THREAD

**Issue by altsst** [soduto/Soduto#30]
https://github.com/sannidhyaroy/Soduto

**Comment by sannidhyaroy** [comment]
My repository is a fork of this official repository. Developers work on existing projects to improve them. There's no such thing as taking over and I don't like the GitHub Issue that you created here. I use the app everyday, it's fantastic and the reason it works is due to the work of the original devs. Not to mention, the devs are still active and there were multiple occasions when they helped me out to implement certain features to the existing codebase.

**Comment by altsst** [comment]
Hey @sannidhyaroy, I just wanted to let people know there is an active fork. I stand by this aswell: "Developers work on existing projects to improve them". Happy to adjust my wording, if you have suggestions.

**Comment by prediscover** [comment]
You should probably just remove the word "abandonded".

**Comment by jace** [comment]
Looks like @sannidhyaroy's fork is also abandoned. Last commit 3 years ago. Another fork by @sidevesh (https://github.com/sidevesh/Soduto) seems the most active, with the last update four months ago. However, there are no releases in it. Yet another downstream fork by @ponzu840w (https://github.com/ponzu840w/Soduto) is the most recently updated (Nov 2025, two months ago) and does have a release (not notarised). This release is working for me.

**Comment by sannidhyaroy** [comment]
Looks like @sidevesh has fixed the compilation issues in the latest Xcode versions. I haven't really worked on the project after joining college as I mostly work on Linux.

I think the project is in better hands though and people are working on it.

**Comment by sannidhyaroy** [comment]
TL;DR: v3.0.0 (24) is out...

---

Okay, I tried pulling all the commits from the downstreams. Go ahead and click on `Menu Bar Icon > Check for Updates`, and hopefully you should receive an update. If not, then you can manually get the release from [here](https://github.com/sannidhyaroy/Soduto/releases/latest/download/Soduto.Nightly.dmg).

Also, the reproducible builds should be working as well. I've dropped support below macOS 11.0 (Big Sur) for reasons explained in the releases page.

Thanks to @sidevesh and @ponzu840w for their immense contributions.

**Comment by sidevesh** [comment]
Hey everyone, the code in my fork is fairly incomplete right now, especially the media controls sync and the input sharing,
my goal with my fork is to support all kdeconnect features on macOS and Android combo and then move beyond kdeconnect in a new fork project,
the code would always remain open source but the repo may change,
I had tried getting some fixes into kdeconnect protocol itself but the changes over there happen very slow and my changes were rejected, hence the decision to eventually fork and go beyond kdeconnect protocols

**Comment by sidevesh** [comment]
In the current state as @sannidhyaroy said I have gotten things to build again,
I was waiting for completing the media controls sync and iron out some bugs before sharing builds,
but for those who can compile themselves and help me fix things :) code is always available

**Comment by sannidhyaroy** [comment]
> I was waiting for completing the media controls sync and iron out some bugs before sharing builds

I shared the same philosophy a year earlier, but then thought that Nightly is meant to be experimental anyway, so I went ahead and released a new Nightly build.

When a certain feature is fairly stable, I'll merge it to beta. Looks like even the basic features like notification icons need tinkering with the KDE Connect Protocol itself, the android app itself isn't sending the icon as payload, so it's a hit or miss.

I'll separate update channels further down the road in Sparkle. I thought of dropping Sparkle in favor of Homebrew, but brew doesn't fulfill the advantages of Sparkle yet and they don't support delta upgrades anyway.

**Comment by sidevesh** [comment]
@sannidhyaroy I would advise not releasing my code at this point as a new release, the work is not complete and there are major issues,
The input sharing can cause a lock situation where the macOS' keyboard mouse inputs get locked out and no way for the user to get control back,
Media controls is also very broken right now, its very hacky and controls don't stay in sync for long and the approach has significant perf hit.
The code is open source so its up to you but right now I don't have the bandwidth either to work on issue reports.

**Comment by sannidhyaroy** [comment]
Hey guys, in an effort to modernize the project, I have tried replacing almost all unmaintained dependencies of Soduto with modern ones that are either built-in or a library that's maintained directly by Apple in their open-source GitHub page (with the exception of Citadel, which is a third-party library built on top of SwiftNIO SSH).

The latest release of `v3.1.0 Nightly` includes the migration of the low-risk dependencies (Reachability.swift and CleanroomLogger). I have started migration of the core dependencies, so could anyone build the project from the `nightly` branch and check if everything is working correctly (including, Soduto Files). It contains the migration of `NMSSH`, `openssl`, `libssh2` and `iSSH2` (managed by Carthage) towards `swift-certificates`+`Citadel` (managed by Swift Package Manager). This simplifies our build tooling as we're no longer depending on C libraries and bridging headers, and also reduces build failure issues on various systems by a lot with the added advantage of not having to patch older abandoned dependencies like `NMSSH` and `iSSH2`.

The downside is dropping support for macOS 11 (Big Sur), 12 (Monterey) and 13 (Ventura). I have carefully analyzed about this in https://github.com/sannidhyaroy/Soduto/issues/45#issuecomment-3861217142 because dropping support for macOS versions (especially after we recently dropped for macOS 10.x and earlier) is an important decision. If that's a problem, please poll and let me know in discussion sannidhyaroy#59

The only remaining unmaintained dependency is `CocoaAsyncSocket` and it's a very critical and core dependency with high risk, so I'm not sure if I'll be successful in replacing it. If we have to keep it, I'll at least migrate it to Swift Package Manager, thus letting us ditch Carthage completely.

Please test the `nightly` branch and let me know how is goes. The Building instructions and Debugging (Logging) instructions are updated, so please re-read them and proceed accordingly.

**Comment by sannidhyaroy** [comment]
`CocoaAsyncSocket` dropped in favor of `SwiftNIO` and Carthage dropped in favor of Swift Package Manager.
All migrations are complete!
Builds should now be dead simple!
Please check the [`nightly`](https://github.com/sannidhyaroy/Soduto/tree/nightly) branch and test if there's no change in end-user functionality.

EDIT: It is advisable for downstreams to pull the changes to avoid accumulating lots of merge conflicts later

@sidevesh can you tell me why you said the media controls were hacky and has a performance hit? From my testing, I didn't find any issue, although I haven't tested much so I maybe wrong. Or, could be that the migrations somehow made it stable. I'd want to take a look at MPRISService next, so I wanted to ask you directly.

**Comment by sidevesh** [comment]
Basically macOS does not allow non system apps to access media playback info anymore,
my implementation uses a library that gets that info through the built in perl binary that still has access to that info,
but this is roundabout, slow and does not provide the complete info, plus Apple may kill this approach too anytime if they want

Plus my implementation was still incomplete, media progress for example was not syncing properly and also we need to add a toggle in the menubar menu to disable this on the fly since if someone uses a dual connection headphone connected to phone and the Mac, the playback being seen on both devices causes them to pause continuously which is really bad,
ideally we should be able to detect these situations but that would have needed going beyond kdeconnect.

Also @sannidhyaroy in the latest release the input sharing plugins in both directions fail because even though accessibility permissions have been provided, I still get the permission pop up and none of them work,
have you seen this issue ?

**Comment by sidevesh** [comment]
SInce you are more actively handling this now, another thing that I was considering was expanding the device name restrictions of the kdeconnect protocol,
it does not allow spaces and inverted commas, which is fine because in Linux land device name is usually hostname,
but in Apple land its not just common but the standard to have names like 'Swapnil's MacBook Pro', which has both those characters and currently they look glitched out on kdeconnect mobile app and cause issues during pairing.

Also kdeconnect has gotten some new plugins recently that again would need support here.

**Comment by sannidhyaroy** [comment]
> SInce you are more actively handling this now, another thing that I was considering was expanding the device name restrictions of the kdeconnect protocol,

I was considering that too, but there's a lot of things to be improved and modernized in Soduto, including device name customizations, complete SwiftUI migration, support for dual sim in TelephonyService hence it'll take time to complete all these. I just added mDNS, pairing verification code and protocol version 8 support in sannidhyaroy/soduto#67

> Also, kdeconnect has gotten some new plugins recently that again would need support here.

Yes, but I'm intentionally keeping them as low priority, because there's a lot of work. I think the digitizer plugin is interesting. Some plugins will never be fully covered unfortunately due to Apple's restrictions, like notifications syncing from macOS to remote device.

EDIT: Also, I think KDE Connect probably has fixed those device name glitches in their latest versions. I can't seem to reproduce it on their latest official app, but Zorin Connect (which didn't pull latest upstream changes) has that glitch. But, yes I'll look into their code, and I do plan to add a feature to customize Soduto's advertised device name.

**Comment by sannidhyaroy** [comment]
> Also @sannidhyaroy in the latest release the input sharing plugins in both directions fail because even though accessibility permissions have been provided, I still get the permission pop up and none of them work,
> have you seen this issue?

Ah no? I just tested it to confirm but Android can control macOS fully (both mouse and keyboard). macOS can control only the mouse of android, not the keyboard but I think either that's not implemented or there's some issue in Soduto itself because no debug logs are generated, so no data packets are sent.

Do you have the accessibility permissions and the remote keyboard enabled in android?

EDIT: I didn't test with the latest release, rather the latest code available in `feat/protocol-v8` branch.

**Comment by sidevesh** [comment]
I got the accessibility issue resolved but had to deep clean and uninstall the old app and all its files after having upgraded from the built in installer and then reinstalling from scratch,
took a few reattempts but then started to work as expected after re-pairing the mobile app.

**Comment by sannidhyaroy** [comment]
> I got the accessibility issue resolved but had to deep clean and uninstall the old app and all its files after having upgraded from the built in installer and then reinstalling from scratch, took a few reattempts but then started to work as expected after re-pairing the mobile app.

No worries, I spent 2 days fighting macOS caching, and another 3 days fighting android's caching, and kept digging around my own code thinking where I went wrong.

Btw, don't we need a space to communicate? There's a lack of communication space, and I faced this badly when I was confused about my decision to raise the minimum version target of Soduto, as I didn't have the community to ask.

There's also a lack of testers, so we will not know when parts of the app breaks unexpectedly. SodutoTests in Xcode is just for testing the certificate flow, and those tests are old and will fail, which defeats the purpose. I also spend almost 6 - 8 hours when publishing a new release. I'm thinking of integrating CI into our workflow, and if we could add some unit tests.

**Comment by sidevesh** [comment]
I think we can create a discord server for the project to coordinate and also we can invite testers there.

**Comment by sannidhyaroy** [comment]
> Basically macOS does not allow non system apps to access media playback info anymore, my implementation uses a library that gets that info through the built in perl binary that still has access to that info, but this is roundabout, slow and does not provide the complete info, plus Apple may kill this approach too anytime if they want

@sidevesh uhh, yes that's correct that macOS restricts access to it's own media playback info for non-system apps. But you haven't implemented reading Mac's local media players and sending them to the remote device. What I have working is macOS showing the media playback of the remote device and showing that in the Now Playing status in the menu bar, and the code in `MPRISService` reflects that. I can't find any perl binary that you mentioned. The code is using proper Apple provided (`MPNowPlayingInfoCenter` / `MPRemoteCommandCenter`) APIs.

EDIT: I've improved the incoming capabilities, and begun adding outgoing capabilities. I understood what you were talking about. The perl script is the only reliable way, we can do this for all apps, other approaches like osascript will work only for limited apps. We can't use `MRMediaRemoteGetNowPlayingInfo` like we used to do since macOS 15.3+, because Apple now requires a private entitlement check (`com.apple.mediaremote.now-playing-read-access`). The perl script works because Apple ships perl as `/usr/bin/perl`, signed by Apple, with bundle ID `com.apple.perl5` and it can load arbitrary native `.dylib`/`.framework` files at runtime. It's not doing any actual work, it's used as a loader because has Apple's trust.
Apple could break this by either:
- Removing `/usr/bin/perl`
- Changing `mediaremoted` to verify the full call chain
- Changing Perl's bundle ID to not start with `com.apple.*`

So, we don't really have a better option anyway tbh.

Sources:
- https://github.com/ejbills/mediaremote-adapter (Swift Package)
- https://github.com/ungive/mediaremote-adapter (CMake library + CLI)
- https://github.com/nohackjustnoobb/media-remote (Rust crate)

**Comment by sannidhyaroy** [comment]
Something nice is cooking for quite a while, guyzz... 😎

Having a downstream client for android lets us ship features faster and improve compatibility down to the protocol-level between android and macOS: https://github.com/sannidhyaroy/soduto-android

As of now, new plugins like LockPlugin, SystemVolumeReceiver and better Notification Icon controls have been added to android, that maps cleanly to macOS.

**Comment by sidevesh** [comment]
@sannidhyaroy are you planning to contribute these changes back to KDE Connect Android ?
I am curious what issues have you faced that led to the fork of android app ?

**Comment by sannidhyaroy** [comment]
> @sannidhyaroy are you planning to contribute these changes back to KDE Connect Android ?
> I am curious what issues have you faced that led to the fork of android app ?

@sidevesh These changes aren't macOS-specific, they're protocol-level improvements that benefit all platforms. For example:

- **Platform and Client identity in device metadata** _(platform name, platform version, client name, client version)_**:** Linux, macOS, Windows, Android and iOS all have meaningfully different capabilities. macOS can't do uinput, it's limited to CGEvent synthesis, and even hardware-level stylus input hits a wall with restricted Apple entitlements (com.apple.developer.hid.virtual.device) that aren't available to unsigned open-source apps. Explicit platform identity lets clients make reliable decisions instead of guessing. Client name and version let both ends know exactly what they're talking to, if both are Soduto, they know the protocol extensions are mutually supported. All extensions are optional, so it degrades gracefully with any other KDE Connect client.

- **Gesture support in Mousepad** _(2/3/4-finger swipe, pinch, rotate)_**:** macOS is heavily gesture-driven, but every modern OS has gestures. The upstream protocol just has no vocabulary for them yet.

- **`canShuffle` / `canLoop` in the MPRIS extension:** Android doesn't support shuffle or loop at all (I added them in the downstream). The clients rely on absence-of-key heuristics to infer support. In our downstream, now it's explicit and deterministic.

- **`canLock` / `canUnlock` in the Lock plugin:** Not all devices can unlock. The protocol should reflect that rather than assume it.

Soduto's userbase is niche and has different expectations from the general KDE Connect audience, so I care a lot about polish and being able to iterate freely. Everything's open-source anyway, so if upstream wants any of it, it's a straightforward pull. I'm open to proposing some of this upstream selectively, though KDE Connect Android is fairly short-staffed and MRs can take years to get through. No fault of anyone, just the reality of a volunteer-driven project.

**Comment by sidevesh** [comment]
hey @sannidhyaroy , how is the update coming along and do have any plans on releasing the new changes soon ?

**Comment by sannidhyaroy** [comment]
> hey @sannidhyaroy , how is the update coming along and do have any plans on releasing the new changes soon ? 

yeah, I've been intending to create a release for months but I've been very busy with my work. I'll try to polish things for release and gather/make changelogs to create a release probably in the a month or so.

**Comment by sannidhyaroy** [comment]
> I think we can create a discord server for the project to coordinate and also we can invite testers there.

Done! [Soduto Labs](https://discord.gg/CEWxVVNHAe) is live on Discord as asked.

This thread has somehow become the coordination hub for the forks: dev discussion, testing reports, release news. All inside the issue tracker of a repo none of us maintain. That's not fair to this project's notification subscribers, and honestly not a great place for us either.

So if you're following any of the active forks, testing builds, or just want to know when something ships: the conversation moves there. Roadmap decisions will get discussed and polled in the open instead of buried in issue comments.

@sidevesh @ponzu840w, see you inside.
